### PR TITLE
Use intra docs links in core::{ascii, option, str, pattern, hash::map}

### PR DESCRIPTION
**Pull request by Amjad50**
Partial fix for #75080

@rustbot modify labels: T-doc A-intra-doc-links

r? @jyn514

### CODE REVIEW

File: library/core/src/str/mod.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@
 //!
 //! For more details, see the [`std::str`] module.
 //!
-//! [`std::str`]: self
+//! [`std::str`]: ../../std/str/index.html
```
Comment (jyn514) [inline library/core/src/str/mod.rs:7]: Why did you revert this link?

File: library/core/src/str/mod.rs [hunk 2]
```diff
@@ -84,8 +84,8 @@ pub trait FromStr: Sized {
     /// when the string is ill-formatted return an error specific to the
     /// inside [`Err`]. The error type is specific to implementation of the trait.
     ///
-    /// [`Ok`]: ../../std/result/enum.Result.html#variant.Ok
-    /// [`Err`]: ../../std/result/enum.Result.html#variant.Err
+    /// [`Ok`]: Result::Ok
+    /// [`Err`]: Result::Err
     ///
```
Comment (jyn514) [inline library/core/src/str/mod.rs:89]: These are in scope from the prelude, so they don't need reference links at all :)
```suggestion
```

File: library/core/src/str/mod.rs (hunk 1 above)
Comment (Amjad50) [reply library/core/src/str/mod.rs:7]: `self` was referring to `core::str`(itself)

Comment (jyn514) [reply library/core/src/str/mod.rs:7]: Oh I see, `std::str` is _not_ a re-export of `core::str`. That's confusing. Good catch!

**Pull request merged**